File: sync-photos.py
#!/usr/bin/env python3
"""
AUTOMATIC LOCAL PHOTO SYNC SCRIPT (Python)
Run with: python3 sync-photos.py
Scans the `images/` directory and indexes any new image files into `data/photos.json`.
"""

import os
import json
import time
from datetime import datetime

CURRENT_DIR = os.path.dirname(os.path.abspath(__file__))
IMAGES_DIR = os.path.join(CURRENT_DIR, 'images')
DATA_FILE = os.path.join(CURRENT_DIR, 'data', 'photos.json')
VALID_EXTENSIONS = {'.jpg', '.jpeg', '.png', '.webp', '.gif', '.avif'}

def sync_photos():
    print(f"🔍 Scanning images folder: {IMAGES_DIR}")

    if not os.path.exists(IMAGES_DIR):
        os.makedirs(IMAGES_DIR, exist_ok=True)

    existing_photos = []
    if os.path.exists(DATA_FILE):
        try:
            with open(DATA_FILE, 'r', encoding='utf-8') as f:
                existing_photos = json.load(f)
                if not isinstance(existing_photos, list):
                    existing_photos = []
        except Exception:
            existing_photos = []

    existing_urls = {p.get('url') for p in existing_photos if isinstance(p, dict)}
    added_count = 0

    for file in sorted(os.listdir(IMAGES_DIR)):
        _, ext = os.path.splitext(file)
        if ext.lower() not in VALID_EXTENSIONS:
            continue

        rel_url = f"images/{file}"
        if rel_url in existing_urls:
            continue

        file_path = os.path.join(IMAGES_DIR, file)
        mtime = os.path.getmtime(file_path)
        date_str = datetime.fromtimestamp(mtime).strftime('%Y-%m-%d')

        clean_name = os.path.splitext(file)[0].replace('-', ' ').replace('_', ' ').title()

        new_entry = {
            "id": f"photo-{int(time.time() * 1000)}",
            "title": clean_name,
            "category": "Random Clicks",
            "url": rel_url,
            "thumbnail": rel_url,
            "date": date_str,
            "description": "",
            "location": "",
            "camera": "",
            "tags": ["photo", "click"]
        }

        existing_photos.insert(0, new_entry)
        existing_urls.add(rel_url)
        added_count += 1
        print(f"➕ Added new photo: {file}")

    if added_count > 0:
        with open(DATA_FILE, 'w', encoding='utf-8') as f:
            json.dump(existing_photos, f, indent=2, ensure_ascii=False)
        print(f"\n🎉 Success! Added {added_count} new photo(s) to data/photos.json.")
    else:
        print("✅ All photos in images/ are already indexed in data/photos.json.")

if __name__ == '__main__':
    sync_photos()

File: sync-photos.js
#!/usr/bin/env node

/**
 * AUTOMATIC LOCAL PHOTO SYNC SCRIPT
 * Run with: node sync-photos.js
 * Scans the `images/` folder for any new image files and adds them to `data/photos.json`.
 */

const fs = require('fs');
const path = require('path');

const IMAGES_DIR = path.join(__dirname, 'images');
const DATA_FILE = path.join(__dirname, 'data', 'photos.json');
const VALID_EXTENSIONS = new Set(['.jpg', '.jpeg', '.png', '.webp', '.gif', '.avif']);

function syncPhotos() {
  console.log('🔍 Scanning images folder:', IMAGES_DIR);

  if (!fs.existsSync(IMAGES_DIR)) {
    fs.mkdirSync(IMAGES_DIR, { recursive: true });
  }

  // Load existing photos.json
  let existingPhotos = [];
  if (fs.existsSync(DATA_FILE)) {
    try {
      const raw = fs.readFileSync(DATA_FILE, 'utf-8');
      existingPhotos = JSON.parse(raw);
      if (!Array.isArray(existingPhotos)) existingPhotos = [];
    } catch (e) {
      console.warn('⚠️ Could not parse existing data/photos.json, creating a fresh one.');
      existingPhotos = [];
    }
  }

  // Set of existing relative URLs
  const existingUrls = new Set(existingPhotos.map(p => p.url));

  // Read files in images/
  const files = fs.readdirSync(IMAGES_DIR);
  let addedCount = 0;

  files.forEach(file => {
    const ext = path.extname(file).toLowerCase();
    if (!VALID_EXTENSIONS.has(ext)) return;

    const relUrl = `images/${file}`;
    if (existingUrls.has(relUrl)) return;

    // Build entry
    const stats = fs.statSync(path.join(IMAGES_DIR, file));
    const title = file
      .replace(/\.[^/.]+$/, '')
      .replace(/[-_]/g, ' ')
      .replace(/\b\w/g, l => l.toUpperCase());

    const dateStr = stats.birthtime ? stats.birthtime.toISOString().split('T')[0] : new Date().toISOString().split('T')[0];

    const newEntry = {
      id: `photo-${Date.now()}-${Math.floor(Math.random() * 1000)}`,
      title: title,
      category: 'Random Clicks',
      url: relUrl,
      thumbnail: relUrl,
      date: dateStr,
      description: '',
      location: '',
      camera: '',
      tags: ['photo', 'click']
    };

    existingPhotos.unshift(newEntry);
    existingUrls.add(relUrl);
    addedCount++;
    console.log(`➕ Added new photo: ${file}`);
  });

  if (addedCount > 0) {
    fs.writeFileSync(DATA_FILE, JSON.stringify(existingPhotos, null, 2), 'utf-8');
    console.log(`\n🎉 Success! Added ${addedCount} new photo(s) to data/photos.json.`);
  } else {
    console.log('✅ All photos in images/ are already indexed in data/photos.json.');
  }
}

syncPhotos();
